xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,441 @@
+    if (xdsClient != null) {
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+    }
+    if (timeService != null) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
+    }
+    if (restartTimer != null) {
+      restartTimer.cancel();
+    }
+    isServing = false;
+    internalTerminationLatch.countDown();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown.get();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    if (internalTerminationLatch.getCount() != 0) {
+      return false;
+    }
+    return delegate.isTerminated();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    long startTime = System.nanoTime();
+    if (!internalTerminationLatch.await(timeout, unit)) {
+      return false;
+    }
+    long remainingTime = unit.toNanos(timeout) - (System.nanoTime() - startTime);
+    return delegate.awaitTermination(remainingTime, TimeUnit.NANOSECONDS);
+  }
+
+  @Override
+  public void awaitTermination() throws InterruptedException {
+    internalTerminationLatch.await();
+    delegate.awaitTermination();
+  }
+
+  @Override
+  public int getPort() {
+    return delegate.getPort();
+  }
+
+  @Override
+  public List<? extends SocketAddress> getListenSockets() {
+    return delegate.getListenSockets();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getServices() {
+    return delegate.getServices();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getImmutableServices() {
+    return delegate.getImmutableServices();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getMutableServices() {
+    return delegate.getMutableServices();
+  }
+
+  // Must run in SynchronizationContext
+  private void startDelegateServer() {
+    if (restartTimer != null && restartTimer.isPending()) {
+      return;
+    }
+    if (isServing) {
+      return;
+    }
+    if (delegate.isShutdown()) {
+      delegate = delegateBuilder.build();
+    }
+    try {
+      delegate.start();
+      listener.onServing();
+      isServing = true;
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(null);
+      }
+    } catch (IOException e) {
+      logger.log(Level.FINE, "Fail to start delegate server: {0}", e);
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(e);
+      }
+      restartTimer = syncContext.schedule(
+        new RestartTask(), RETRY_DELAY_NANOS, TimeUnit.NANOSECONDS, timeService);
+    }
+  }
+
+  private final class RestartTask implements Runnable {
+    @Override
+    public void run() {
+      startDelegateServer();
+    }
+  }
+
+  @VisibleForTesting
+  void setTimeService(@Nullable ScheduledExecutorService scheduledExecutorService) {
+    if (scheduledExecutorService != null) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
+    }
+    timeService = scheduledExecutorService;
```

Comment:
if `scheduledExecutorService` is null you are going to set `timeService` also to null and lose the old value and not release it?